### PR TITLE
Add support for array first/last dot notation

### PR DESCRIPTION
**Pull request by funkjedi**
- [x] I've run the tests with `vendor/bin/phpunit`
- [x] None of the tests were found failing
- [x] I've seen the coverage report at `build/coverage/index.html`
- [x] Not a single line left uncovered by tests
- [x] Any coding standards issues were fixed with `vendor/bin/php-cs-fixer fix`

> Array first and last. If you have an expression whose value is an array, you can follow it with .first or .last to resolve to its first or last element.
https://github.com/Shopify/liquid/wiki/Liquid-for-Designers

### CODE REVIEW

File: src/Liquid/Context.php [hunk 1]
```diff
@@ -330,6 +330,16 @@ private function variable($key)
 			}
 
 			if (is_array($object)) {
+				// if the last part of the context variable is .first we return the first array element
+				if ($nextPartName == 'first' && count($parts) == 0 && !Liquid::arrayIsAssoc($object)) {
+					return StandardFilters::first($object);
+				}
+
+				// if the last part of the context variable is .first we return the last array element
```
Comment (sanmai) [inline src/Liquid/Context.php]:
```suggestion
				// if the last part of the context variable is .last we return the last array element
```

File: src/Liquid/Liquid.php [hunk 2]
```diff
@@ -151,4 +151,18 @@ public static function arrayFlatten($array)
...
 	}
+
+	/**
+	 * Check if an array is associative or not.
+	 *
+	 * @param array $array
+	 *
+	 * @return bool
+	 */
+	public static function arrayIsAssoc($array)
```
Comment (sanmai) [inline src/Liquid/Liquid.php]:
```suggestion
	public static function arrayIsAssoc(array $array)
```

**Comment by sanmai** [comment]
Why do we avoid getting the first/last values of an associative arrays? 

Standard filters work with them too.

**Comment by funkjedi** [comment]
Nothing I'm married to here's the thought processes.

1. The documentation seems to make an explicit distinction between arrays and hashes within Liquid and that first/last were only for arrays so I just went with that.

2. Could potentially introduce breaking changes for any associative arrays with `first` or `last` keys. For example the `forloop` variable added by the *For* tag.

**Comment by sanmai** [comment]
PHP-wise that's kind of surprising. You can assume if you have an array, you can get the last/first elements. What we're losing if we chose to go the PHP-way, and treat all arrays the same?

I assume that tests won't break. If they are, there's no question.

**Comment by sanmai** [comment]
There's no distinction for arrays/hashes in PHP, and the documentation (from the screenshot) does not prohibit calling `.first` or `.last` on arrays, as I see it. Yes, it will be backward-incompatible in a sense that you won't be able use the same template under Ruby's Liquid, but that's the price to pay. How do you think?

**Comment by sanmai** [comment]
Thank you for handling associative arrays. Tests similar to `testOverrideSize` have to be added then.

**Pull request merged**